File: modules/expo-thumbhash/ios/ExpoThumbHashModule.swift
import Foundation
import ExpoModulesCore

public class ExpoThumbHashModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoThumbHash")

    // MARK: - Functions

    AsyncFunction("decodeToDataURIAsync") { (base64Hash: String, promise: Promise) in
      // Decode on background thread — PNG encoding is expensive
      DispatchQueue.global(qos: .userInitiated).async {
        guard let image = ThumbHashDecoder.shared.decode(base64Hash: base64Hash),
              let pngData = image.pngData()
        else {
          promise.resolve(nil as String?)
          return
        }
        promise.resolve("data:image/png;base64," + pngData.base64EncodedString())
      }
    }

    Function("approximateDimensions") { (base64Hash: String) -> [String: Int] in
      let (w, h) = ThumbHashDecoder.shared.approximateDimensions(base64Hash: base64Hash)
      return ["width": w, "height": h]
    }

    Function("clearCache") {
      ThumbHashDecoder.shared.clearCache()
    }

    // MARK: - View

    View(ThumbHashView.self) {
      Prop("thumbHash") { (view: ThumbHashView, prop: String) in
        view.thumbHash = prop
      }

      Prop("crossFadeDuration") { (view: ThumbHashView, prop: Double) in
        view.crossFadeDuration = prop / 1000.0 // ms → seconds
      }
    }
  }
}

File: modules/expo-thumbhash/android/src/main/java/expo/modules/thumbhash/ExpoThumbHashModule.kt
package expo.modules.thumbhash

import android.util.Base64
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoThumbHashModule : Module() {
  override fun definition() =
    ModuleDefinition {
      Name("ExpoThumbHash")

      AsyncFunction("decodeToDataURIAsync") { base64Hash: String ->
        val bitmap = ThumbHashDecoder.decode(base64Hash) ?: return@AsyncFunction null

        val stream = java.io.ByteArrayOutputStream()
        bitmap.compress(android.graphics.Bitmap.CompressFormat.PNG, 100, stream)
        val pngBytes = stream.toByteArray()
        val pngBase64 = Base64.encodeToString(pngBytes, Base64.NO_WRAP)

        return@AsyncFunction "data:image/png;base64,$pngBase64"
      }

      Function("approximateDimensions") { base64Hash: String ->
        val (w, h) = ThumbHashDecoder.approximateDimensions(base64Hash)
        return@Function mapOf("width" to w, "height" to h)
      }

      Function("clearCache") {
        ThumbHashDecoder.clearCache()
      }

      View(ThumbHashView::class) {
        Prop("thumbHash") { view: ThumbHashView, hash: String ->
          view.thumbHash = hash
        }

        Prop("crossFadeDuration") { view: ThumbHashView, duration: Int ->
          view.crossFadeDuration = duration
        }
      }
    }
}
